includes/enrolment/class-sensei-enrolment-provider-journal-store.php

```diff
@@ -0,0 +1,283 @@
+	}
+
+	/**
+	 * Register a possible update in the status of a provider for a user and a course. If there was no actual change on
+	 * the status, this method has no effect.
+	 *
+	 * @param array $provider_results An array with the format 'provider_id' => enrollment result.
+	 * @param int   $user_id          The user which the change applies to.
+	 * @param int   $course_id        The course which the change applies to.
+	 */
+	public static function register_possible_enrolment_change( $provider_results, $user_id, $course_id ) {
+		$journal_store = self::get( $user_id, $course_id );
+
+		$has_changed = false;
+
+		foreach ( $provider_results as $provider_id => $is_enrolled ) {
+			if ( ! isset( $journal_store->providers_journal[ $provider_id ] ) ) {
+				$journal_store->providers_journal[ $provider_id ] = Sensei_Enrolment_Provider_Journal::create();
+			}
+
+			$has_changed = $journal_store->providers_journal[ $provider_id ]->update_enrolment_status( $is_enrolled ) || $has_changed;
+		}
+
+		$current_snapshot = self::get_enrolment_snanpshot( $user_id, $course_id );
+
+		$removed_providers = array_diff( array_keys( $current_snapshot ), array_keys( $provider_results ) );
+		foreach ( $removed_providers as $removed_provider ) {
+			$has_changed = $journal_store->providers_journal[ $removed_provider ]->delete_enrolment_status() || $has_changed;
+		}
+
+		$journal_store->has_changed = $has_changed;
+	}
+
+	/**
+	 * Get a snapshot of the providers' enrolment status for a user and course at a specific timestamp. If no timestamp
+	 * is provided the current snapshot is returned.
+	 *
+	 * @param int       $user_id     The user to return the snapshot for.
+	 * @param int       $course_id   The course to return the snapshot for.
+	 * @param int|float $timestamp   The timestamp of the snapshot. If omitted the current snapshot is returned.
+	 *
+	 * @return array
+	 */
+	public static function get_enrolment_snanpshot( $user_id, $course_id, $timestamp = null ) {
+		$timestamp     = null === $timestamp ? microtime( true ) : (float) $timestamp;
+		$journal_store = self::get( $user_id, $course_id );
+
+		$snapshot = [];
+		foreach ( $journal_store->providers_journal as $provider_id => $journal ) {
+			$status = $journal->get_status_at( $timestamp );
+
+			if ( null !== $status['enrolment_status'] ) {
+				$snapshot[ $provider_id ] = $status['enrolment_status'];
+			}
+		}
+
+		return $snapshot;
+	}
+
+	/**
+	 * Get the enrolment status history of a provider for a specified user and course.
+	 *
+	 * @param int $user_id     The user id.
+	 * @param int $course_id   The course id.
+	 * @param int $provider_id The provider id.
+	 *
+	 * @return array The history of the provider. Each element of the array has the format:
+	 *               [ 'timestamp' => Timestamp of the status change, 'enrolment_status' => true|false|null ]
+	 */
+	public static function get_provider_history( $user_id, $course_id, $provider_id ) {
```

Comment:
Should we have this one match below?

Reply:
Missed that one! cb743eadf2680930d9f80b5167dc063817d5d686